# SnakeBin
<img width="2000" height="758" alt="immagine" src="https://github.com/user-attachments/assets/a98b03c8-09f3-42b3-b343-6bfae4947a01" />

Repository developed as part of my PhD thesis. 
SnakeBin is a Snakemake pipeline for insert-size-resolved Whole Exome Sequencing (WES) analysis of NGS data, stratifying reads into normalized Insert Size Bins (ISBs) to quantify fragment length effects on coverage efficiency, mappability, genotypability, and variant calling.

Starting from raw fastq files, it aligns them against the reference genome and extracts per-read insert sizes (TLEN), bins fragments (e.g., 50bp intervals), and computes ISB-stratified metrics including depth/bredth/uniformity of coverage, on/near/off-target fractions, mapping quality histograms, duplication rates, callable bases (as defined by GATK CallableLoci).

- Coverage-Based Downsampling 
  - Merges sample's fastq, trims and aligns reads, generates ISBs, subsamples reads per ISB to fixed target depth across bins (e.g., 50x average coverage, eliminating depth confounding effects), postprocesses BAMs, computes relevant metrics and optional variant calling.

- Fragment-Based Downsampling 
  - Aligns samples, generates sample-specific ISBs, converts BAMs to FASTQs, sequentially subsamples ISBs by selecting the first nÂ° fragments (ensuring an equal contribution of each sample), merges sample-specific ISBs to cohort-specific ISBs, realigns the ISBs, postprocesses BAMs, computes relevant metrics and optional variant calling.
â€‹

---

## ğŸš€ Features
- Insert-size resolved WES Analysis
- Fragment-based normalization
- Coverage-based normalization
- Variant calling
- Scalability
---

## ğŸ“ Repository Structure

SnakeBin \
â”œâ”€â”€ config # config YAML file \
â”œâ”€â”€ workflow \
â”‚ â”œâ”€â”€ SnakeFile # workflow \
â”‚ â”œâ”€â”€ rules \
â”‚ â”‚ â”œâ”€â”€ downsampling # snakefiles with the downsampling rules \
â”‚ â”‚ â”œâ”€â”€ fastqc # snakefile with fastqc rules \
â”‚ â”‚ â”œâ”€â”€ mapping #snakefiles with mapping, base recalibration, clipping, duplicate removal and flagstat rules \
â”‚ â”‚ â”œâ”€â”€ merging # snakefiles with the fastq merging rules \
â”‚ â”‚ â”œâ”€â”€ positionbaseddownsampling # snakefiles with the picard position based 1% fragment downsampling with the relative duplicate removal and flagstat rules \
â”‚ â”‚ â”œâ”€â”€ positionbaseddownsampling_per_sample # like above, but per-sample \
â”‚ â”‚ â”œâ”€â”€ snv_calling # snakefiles with the vairant calling and filtering rules \
â”‚ â”‚ â”œâ”€â”€ split_bin # snakefiles with the insert size bin BAM generation rules \
â”‚ â”‚ â”œâ”€â”€ stats # snakefiles with the metrics computation rules \
â”‚ â”‚ â”œâ”€â”€ trimming # snakefiles with the trimming rules \
â”‚ â”œâ”€â”€ scripts # additional python scripts for the metrics collection \
â””â”€â”€ README.md


---

## ğŸ› ï¸ Installation and Usage

```bash
#Clone locally the repository: 
git clone https://github.com/Lab-Delledonne-bioinfo/SnakeBin
cd SnakeBin

â–¶ï¸ Usage
#If not installed globally, activate the snakemake conda environment:
conda activate snakemake

#Example of pipeline testing in dry-run mode:
snakemake --cores 1 --dry-run

#Example of pipeline running in nohup:
nohup snakemake --cores 10 > Example_Snakemake.log &

#Temporary data cleanup after pipeline computation:
snakemake --delete-temp-output #cleanup temp files
```
âš™ï¸ Configuration

#Before running the pipeline edit following file. In particular, edit the sample/result folder path, insert size binning scheme, downsampling levels, target regions files for metrics computation, tools executable files.
```
config.yaml
```

ğŸ“¦ Dependencies
```bash
#Orchestration
Snakemake 
#main language
Python 3, java 1.8, java (latest)
#Libraries (python)
os, sys, collections, gzip, glob, csv, pandas, datetime, re, subprocess, shutil
#Tools
fastqc, fastp, BWA-MEM2, samtools, picard, bamutil, bedtools, GATK v3.8, GATK (latest), bcftools, RTG tools, sambamba
```

ğŸ‘¤ Author \
Email: lucabertoli10@yahoo.it








